external/storm-kafka-monitor/src/main/java/org/apache/storm/kafka/monitor/KafkaOffsetLagUtil.java

```diff
@@ -0,0 +1,374 @@
+     * @return log head offset, spout offset and lag for each partition
+     */
+    public static List<KafkaOffsetLagResult> getOffsetLags (OldKafkaSpoutOffsetQuery oldKafkaSpoutOffsetQuery) throws Exception {
+        List<KafkaOffsetLagResult> result = new ArrayList<>();
+        Map<String, List<TopicPartition>> leaders = getLeadersAndTopicPartitions(oldKafkaSpoutOffsetQuery);
+        if (leaders != null) {
+            Map<String, Map<Integer, Long>> logHeadOffsets = getLogHeadOffsets(leaders);
+            Map<String, List<Integer>> topicPartitions = new HashMap<>();
+            for (Map.Entry<String, List<TopicPartition>> entry: leaders.entrySet()) {
+                for (TopicPartition topicPartition: entry.getValue()) {
+                    if (!topicPartitions.containsKey(topicPartition.topic())) {
+                        topicPartitions.put(topicPartition.topic(), new ArrayList<Integer>());
+                    }
+                    topicPartitions.get(topicPartition.topic()).add(topicPartition.partition());
+                }
+            }
+            Map<String, Map<Integer, Long>> oldConsumerOffsets = getOldConsumerOffsetsFromZk(topicPartitions, oldKafkaSpoutOffsetQuery);
+            for (Map.Entry<String, Map<Integer, Long>> topicOffsets: logHeadOffsets.entrySet()) {
+                for (Map.Entry<Integer, Long> partitionOffsets: topicOffsets.getValue().entrySet()) {
+                    Long consumerCommittedOffset = oldConsumerOffsets.get(topicOffsets.getKey()) != null ? oldConsumerOffsets.get(topicOffsets.getKey()).get
+                            (partitionOffsets.getKey()) : -1;
+                    consumerCommittedOffset = (consumerCommittedOffset == null ? -1 : consumerCommittedOffset);
+                    KafkaOffsetLagResult kafkaOffsetLagResult = new KafkaOffsetLagResult(topicOffsets.getKey(), partitionOffsets.getKey(),
+                            consumerCommittedOffset, partitionOffsets.getValue());
+                    result.add(kafkaOffsetLagResult);
+                }
+            }
+        }
+        return result;
+    }
+
+    private static Map<String, List<TopicPartition>> getLeadersAndTopicPartitions (OldKafkaSpoutOffsetQuery oldKafkaSpoutOffsetQuery) throws Exception {
+        Map<String, List<TopicPartition>> result = new HashMap<>();
+        if (oldKafkaSpoutOffsetQuery.getPartitions() != null) {
+            String[] partitions = oldKafkaSpoutOffsetQuery.getPartitions().split(",");
+            String[] leaders = oldKafkaSpoutOffsetQuery.getLeaders().split(",");
+            for (int i = 0; i < leaders.length; ++i) {
+                if (!result.containsKey(leaders[i])) {
+                    result.put(leaders[i], new ArrayList<TopicPartition>());
+                }
+                result.get(leaders[i]).add(new TopicPartition(oldKafkaSpoutOffsetQuery.getTopic(), Integer.parseInt(partitions[i])));
+            }
+        } else {
```

Comment:
when can this scenario occur? There is a lot of code in some of these methods. It would be ideal to put some comments hinting on what is going on, and/or split into smaller methods.